README.md
[<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/a/ae/Flag_of_the_United_Kingdom.svg/100px-Flag_of_the_United_Kingdom.svg.png" height="30">](README.md)
[<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/a/a4/Flag_of_the_United_States.svg/100px-Flag_of_the_United_States.svg.png" height="30">](README.md)
[<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/b/ba/Flag_of_Germany.svg/100px-Flag_of_Germany.svg.png" height="30">](README.de_DE.md)

# Inofficial modification of the RF firmware for RF1000 and RF2000 devices
Based on Repetier-Firmware - the fast and user friendly firmware.

## Disclaimer

THIS VERSION IS AN UNOFFICIAL MODIFICATION OF mhiers UNOFFICIAL MODIFICATION OF THE ORIGINAL FIRMWARE. It is neither endorsed nor
supported by the developers of the RF1000 firmware.  
USE AT YOUR OWN RISK.

For the official version, have a look at the upstream repository:
https://github.com/RF1000/Repetier-Firmware (see branch development)

## HowTo Install

- Download the Firmware `Branch: community_development` and unzip all the files.  
- Install Arduino.cc 1.6.5 or 1.8.4 and later if it is not installed on your computer already.  
- Edit and save Configuration.h @Line46 and 47 according to your printers model. You have to remove the two **//** in front of the printers name you wish to activate:  
`#define MOTHERBOARD                         DEVICE_TYPE_RF1000` or  
`#define MOTHERBOARD                         DEVICE_TYPE_RF2000`
- Optional: Edit and save the Configuration.h and RF1000.h or RF2000.h to your needs in case you use special specialized hardware.
- Connect your Printer with a USB-cable and switch it on.
- Go to and doubleclick /Repetier/Repetier.ino
- Within the cyan colored Arduino-IDE you have to choose some preconfiguration settings for your RFx000 Board at the menu **Tools->**:  
Board: `Arduino Mega 2560 or Mega ADK`,  
Processor: `ATMega 2560 (Mega 2560)`,  
Port: Your printers Port like `COM3`, if successfully connected.
- **Check** and **Upload** the Firmware to your Printer.
## If you upgrade to this Version from 1.37r or earlier please do a fresh M303 PID-Autotune on all heaters!  
- RF2000/RF1000 extruder 1: `M303 P0 X0 S230 R10 J1`
- RF2000 extruder 2: `M303 P1 X0 S230 R10 J1`
- RF2000 heated bed: `M303 P2 X0 S70 R15 J3`
- RF1000 heated bed: `M303 P1 X0 S70 R15 J3`  
And start with EEPROM-values `PID drive min = 5` and `PID drive max = 100`  
Or set #define PID_CONTROL_DRIVE_MIN_LIMIT_FACTOR to 10.0f in configuration.h to use the old version PID-control.

## Version RF.01.37mod 

See:  
http://www.rf1000.de/viewtopic.php?f=74&t=1674 (Nibbels/Wessix SenseOffset-Thread)  
http://www.rf1000.de/viewtopic.php?f=7&t=1504#p14882 (mhier Mod)  
http://www.rf1000.de/viewtopic.php?f=7&t=1504&start=60#p16397 (added feature)  
http://www.rf1000.de/viewtopic.php?f=7&t=1610 (rf1000s developement version thread to 1.35)  
http://www.rf1000.de/viewtopic.php?f=7&t=1665 (rf1000s developement version thread to 1.37)  
https://github.com/Nibbels/Repetier-Firmware/commits/heat_bed_z_offset_scan-%26-heat_bed_decrease (all the commits)

## Main Features of this Mod:
* mhier Z-Offset-Scan
* Nibbels/Wessix SensiblePressure Adjustment

## List of Features and additional G-Codes for RF2000 and RF1000

_by mhier_:  
M3900               - Scan the heatbeds offset and correct the active zMatrix within the RAM of the Printer. (X-Y-Pos, Learning-Factor, linear distance weight.)  
RF2000: Z_OVERRIDE_MAX = 1.0mm, instead of 0.5mm  
RF1000: Z_OVERRIDE_MAX = 1.0mm, instead of 0.5mm  

_by StarTony_:  
Fan-Speed-Patch, see http://www.rf1000.de/viewtopic.php?f=7&t=1638  

_by Wessix_:  
M355 Sx - Turn on and off, or switch Port X19 to controll lights etc. (24V-MOSFET @Connector X19)  
M355    - Switch Port X19 to controll lights etc. (24V-MOSFET @Connector X19)  
Idea to implement the SensiblePressure-Function, see M3909  
Idea to send the active digits with the printTemperature-Status.

_by Nibbels_:  
M3900 Xn Ym Px Sy   - Scan the heatbeds offset and correct the active zMatrix within the RAM of the Printer. (X-Y-Pos, Learning-Factor, linear distance weight.)  
M3901 Xn Ym Px Sy   - to preconfigure M3900 (X-Y-Pos, Learning-Factor, linear distance weight.)  

M3902 R1            - to fix a single HeatBeds Hole within the HBS Matrix.  
M3902 Zn.n          - to add an Offset to the Matrix. n.n = {-0.2 .. 0.2} [mm]  
M3902 Z0            - to shift your active zOffset to the zMatrix. The M3006 zOffset will be zero afterwards. The zMatrix is altered within Ram and might be saved to EEPROM[n] with M3902 S[n] afterwards.  
M3902 Sn            - to save the active Matrix to position n = {1..9}  
M3902 Z0 S1         - to shift the zOffset to your zMatrix and save the Matrix at position 1. This is an example to show that the options of M3902 can be combined.  
M3939 Fn St1 Pt2 Ex Iy Rm   - to messure a curve of your Filaments velocity over digits -> viscosity.  

M3920 Sb - to go into or switch back from SilentMode (This will lower your steppers current to an alternative profile)  
Removed all compiler warnings and errors within the original Firmware.  
Included some Fixes the original developers of the branch "repetier/Repetier-Firmware" committed to their firmware.  
Upgraded the Firmware to the latest RF.01.37 (2017-01-20)  
Activated the X35 Temperature-Sensor for RF2000.  
Listed all spare pins for RF1000 or RF2000.  
Fixed the RF2000's Status Message length to 20chars.  
Fix for M3117
Third Z-Scale Mode: G-Code @ Configuration->Z-Configuration->Z-Scale.

_by Nibbels/Wessix_:  
M3909 Pn Sm         - See "Nibbels/Wessix SenseOffset"  

## mhier Z-Offset-Scan
* M3900 Xn Ym Sy Px             - Run a Z-Offset Scan at the specified Location.

Preconfiguration Options for M3900:  
* M3901 Xn Ym Sy Px 

[X]/[Y] specifies the Location of the Scan within the original HBS-Locations.  
X = {1..9}, {0 = random}
Y = {2..12}, {0 = random}
With two extruders X=1 is allowed otherwise forbidden. Normally X=2 is the lowest possible scan position. This limit is caused by border values within every zMatrix. You cannot set it wrong, the upper and lower limits are limited and autoadjusted by the firmware if set wrong.  

[S] specifies the learning rate of M3900  
S= {100}  
The measured offset will be added/substracted by 100% of its Value from/to the original matrix. The original matrix is the matrix stored in your EEPROM. Thus the matrix will be reloaded from the EEPROM to the RAM prior to the following z-offset scan.
S= {0..99}  
the Matrix in RAM will be adjusted by 0% to 99% of the messured offset. This is a great feature for multiple incremental corrections, in case you would not want to start allover or you cannot trust your first values 100% (you want to sum up corrections from different scanning-locations).

[P] ZOS learning linear distance weight  
P = {0..100}  
This is a configuration option to commit a scans offset to its surrounding area only.  
If you scan the middle of the heated bed and then all the corners with a high distance weight, this feature will behave like autobed leveling of your zMatrix.
Example: Put P to a high value and messure all corners. Then the matrix is somehow "bended" and not constantly updated. (Might work very well on heat-beds which tend to lift edges within different temperature ranges or to correct tilting).

You can activate the mhier Z-Offset-Scan within ->Configuration->Z-Configuration->Z-Offset Scan within your printers menu structure. This will always use the preconfigured or changed (see M3901) settings within your printers RAM. These M3900/M3901-Settings are not stored within the printers EEPROM.  

## Nibbels/Wessix SenseOffset  
* M3909 Pn Sm  
P = max. digits = {1...14999} [digits]  
S = max. SenseOffset = {1...300} std.: 180 [um]  

Feature called "Nibbels/Wessix SenseOffset". This is a digit limiter which decreases the printers digits by lowering the heatbed. It compensates the thermal expansion of some second order printer parts.  
 The printer will automaticly release Pressure inbetween the Nozzle and the HeatBed, whenever the digits rise atop [max.digits] by adjusting the Z-Offset. 
 The adjustment-offset is fixed to positive values (bed does never go closer to the nozzle than without M3909). The adjustment will not violate the "max. SenseOffset" restriction in order to avoid unwanted offsets (not caused by too close nozzle).  
 When choosing "max. SenseOffset" you should already know how much digits your printer normally has (Same Material, same Temperatures, same Speed, same Nozzle) and then add some plus 20% or plus 1000digits of force-tolerance.  
 Senseoffset can have two states. Active controling and idle. It is idle, whenever Z is not within negative and ~0.3mm height. You can tune this height limit by tweaking M3007 (http://www.rf1000.de/wiki/index.php/GCodes#M3007_-_Minimale_H.C3.B6he_f.C3.BCr_Z-Kompensation_festlegen). Whenever there is active control you see a circumflex "^" within the modmenu, instead of the @-Symbol.
 ![ ](https://downfight.de/picproxy.php?url=https://downfight.de/grafiken/dimi/modmenudisplay_eng.png "Angebratenes ABS")  
 When the z-compensation gets deactivated the senseoffset-function is deactivated as well.  
 Use "M3909 P0" for manual shutdown of the feature, but normally this is not necessary.  
 You cannot activate the Feature if zCompensation is not active already.  
 
 While Printing you can adjust or activate this feature with the "Left"- and "Right"-Button, while being in the "Mod-Menu" which is page 2 on the printers display.  

## SensibleViscosity  
* M3939 Fn St1 Pt2   Ex Iy Rm  
S = Starting Temperature for the Test in [°C]  
P = Final Temperature for the Test in [°C]  
F = max. test digits = 10000 or {1...12000} [digits]  
E = max. test extrusion velocity = 5 or {0.05...maxStartFeedrate} [mm/s]  
optional:  
I = test velocity increment = 0.1 or {0.02...0.4} [mm/s]  
R = max. digit change needed to consider having a filled nozzle. = 800 [digits] 

Feature called "SensibleViscosity"  
 The printer will automaticly extrude into the thin air and messure the resulting force. 
 After every messurement the velocity of extrusion is increased.  
 You will get a CSV-Output within your Console-Window (Repetierserver/Repetierhost/Octoprint/...) which you can easily copy and paste to some text file. Excel can then read the Data and you can put it into some Chart with 2 Clicks.
 You might ask why you need this feature, but in case you know your Filament quite well you can compare these resulting chars to really see how your filament extrusion behaves under different circumstances or when your nozzle is or gets clogged.
 For longer heated/boiled (my) ABS: http://www.rf1000.de/viewtopic.php?f=62&t=1549&p=16617#p16617  
 ![ ](https://downfight.de/picproxy.php?url=http://image.prntscr.com/image/3508699873f7431489d1df9344110e7f.png "Boiled ABS")  
 For the general viscosity of (my) ABS: http://www.rf1000.de/viewtopic.php?f=23&t=1620&p=16613#p16605  
 ![ ](https://downfight.de/picproxy.php?url=http://image.prntscr.com/image/610de7720945474590668b20c05d7652.png "Temperature ABS")  
 That is a clogged nozzle with the same ABS:
 ![ ](https://downfight.de/picproxy.php?url=http://image.prntscr.com/image/c00ac11b3a384994b37ce8bc3cf03bd9.png "Clogged Nozzle ABS")  
 That is red PLA:
 ![ ](https://downfight.de/picproxy.php?url=http://image.prntscr.com/image/2a3253c930794afc81e4fa4d4b2a4261.png "Clogged Nozzle ABS")  

 Have fun finally seeing your Filaments behaviour over time, temp, nozzles ;)  
 
## Configurable emergency zStop and emergency pause digits within menu and EEPROM  
Menü -> Configuration -> General  

## Configurable motorcurrent within menu and EEPROM  
You can easily change the printers stepper currents within menu for more silence. This features is a replacement to our old silence-feature by gcode. EEPROM support included. The settings are restored after a reboot.
Menü -> Configuration -> Stepper  

## Configurable temperature settings within menu and EEPROM  
PID-Autotune, other controlparameters and the sensortype is adjustable within the menu.  
Menü -> Configuration -> Temperatures -> Extruder0/Extruder1/Heizbett  

## Dualhead Tip-Down Support (beta)
* M3919 [S]mikrometer - Testfunction for Dip-Down-Hotend @ T1: T1 can now be springloaded and the bed will be adjusted "down" whenever T1 is selected. This is alike Ultimaker 3 does it with the right hotend.   
Example: M3919 T1 Z-0.6 tells the Printer that the right hotend will reach 0.6mm more down than the left hotend. Test it :)
* EEPROM Support for offset

## Feature Digit-Z-Kompensation (beta)
If you apply force to the DMS sensors the bed is adjusted (if Z-Compensation is active). This corrects the error which bending of the sensors apply to your layers. The physical change of the beds level is approximately +0,01mm for +1000 applied force (digits).

## RF2000: Additional Temperature Sensor
This optional 3rd temperature T3 is automatically sent out with the other Temperatures and Digits.  
``` 15:16:14.637: T:204.67 /205 B:28.60 /20 B@:0 @:143 T0:28.60 /0 @0:0 T1:204.67 /205 @1:143 T3:28.23 F:322 ```
See RESERVE_ANALOG_SENSOR_TYPE in RF2000.h for configuration options.  
See RESERVE_ANALOG_INPUTS to find the involved code within the firmware.  
You will need to install and additional cable and a temperature sensor at X35 on your printers Board. Then you can wire your sensor to some place needed.  
This sensor might keep track of your boards temperature or the air-temperature within your enclosure.  

## Advanced profiles for PID Autotune

Additions to GCode M303:
[J0] Classic PID  
[J1] Pessen Integral Rule (suggested for Hotend)  
[J2] Some Overshoot  
[J3] No Overshoot (suggested for Heated Bed)  
[J4] until [J6] are PD, PI, P-Profiles for control experts and special cases.  
[Rn] Configurable Autotune cycles  
Easy autotune support via printers menu.  
Siehe auch http://www.rf1000.de/viewtopic.php?f=7&t=1963  

## advanced Pause while printing
Temperature drop while paused.  
Lower motor current on both extruders while paused to prevent unnecessary heat to the extruder and filament.  
Manual extrusion when paused resets automatic retract for continue.

## Wessix`s help video:
[![ScreenShot](https://downfight.de/picproxy.php?url=http://image.prntscr.com/image/d7b7fade0c7343eeb67b680339478894.png)](http://youtu.be/iu9Nft7SXD8)

## !! 22.08.2017: Project is Work in Progress and untested changes are possible.

changelog.txt
V RF.01.37s9.DGMod (2017-08-10)
- Pushing digibike parameters to this branch.

V RF.01.37s9.Mod (2017-08-10)
- Z-Max Fix

V RF.01.37s7.Mod (2017-08-10)
- Optional Temperature Sensor on RF2000 now reports the Temperatures as Heated-Chamber Sensor. https://forum.repetier.com/discussion/4102/heated-chamber -> Letter C in printTemperatures. This is fun, because Repetier-Server 0.86.2 now has an extra temperatures diagram for this.
Activate heated chamber temp diagramm in printer settings -> extruder -> heated chamber -> Yes. Of course RF2000 only has a sensor and no heater/control!

V RF.01.37s6.Mod (2017-08-09)
- found the double-pause-bug. It should be gone now. :D
- change, sort and rewrote the pause functions
- changed Emergency-Pause to double Pause.
- removed warning when changed position 1mm/10mm/50mm multiple times and too fast.
- cleaned up "StepperPositions"
- added ADVANCE to DirectSteps
- ADVANCE increment rises with high values

V RF.01.37s5.Mod (2017-08-08)
- When DirectDrive was activ in Z sometimes the last State of "unfinished direct-Z-driving" was still set. Then the Z-Compensation did not work. I am now trying to fix or override this.
If you still experience problems with first layer tell me!
- Increased ADVANCE factor L Limit to 100 within Menu (was 1..60). Btw: EEPROM-Support has no limit rightnow. And ADVANCE seems to be awsome :D

V RF.01.37s4.Mod (2017-08-07)
- Fixed E1 motor current menu
- Removed Feature Emergency-Stop after 3x Pause

V RF.01.37s3.Mod (2017-08-06)
- Patched ADVANCE Feature to newest Repetier development version.

V RF.01.37s2.Mod (2017-08-06)
- Pulling Repetiers new Guardian-Patch for Timer1 from https://github.com/repetier/Repetier-Firmware/commit/9c2d7323f84f07b33450483b131f4b65d77eb2b7

V RF.01.37s.Mod (2017-08-06)
- Factor L for lineare advance-feature is now available within menu @Configuration->Stepper->Ex-Advance L:
See also:
https://github.com/repetier/Repetier-Firmware/wiki/Hardware-settings-and-print-quality
http://www.dr-henschke.de/advance.html
http://reprap.org/pipermail/reprap-dev/2011-May/003323.html
- Sensortype 1 (Pico Hotend, Reprap style Thermistor) EPCOS NTC is now accessible through menu for all Extruders
- Sensortype 4 is now available through menu for alternative Heated Bed Silicon-Heater. (> http://www.rf1000.de/viewtopic.php?f=29&t=1364&p=13632&hilit=Ceran+Eigenbau#p13632 )
- Small cleanup on several locations

V RF.01.37r9.Mod (2017-08-04)
- Added Yes/No Question to restore configuration defaults menu
- appended missing values to restore configuration function
- Added Steps/mm to stepper menu
- Finally fixed second Extruder "I-Drive MAX Configuration Bug" @RF2000

V RF.01.37r8.Mod (2017-08-04)
- Added Menu for Motor-Current Configuration
- Added EEPROM_Support for EEPROM-Support
- Added second Extruder to "Pause-Current-Support of Pause-Feature" (which did not work on RF2000 or Dual-RF1000)
- Renamed MOTOR_CURRENT to MOTOR_CURRENT_MAX
- Renamed MOTOR_CURRENT_SILENT to MOTOR_CURRENT_NORMAL
- Increased standard motorcurrent for Z to 95 instead of 90
- Depreached/Removed Mod-Feature Silent_Mode because the new settings are better.

V RF.01.37r7.Mod (2017-08-03)
- Added Menu for PID Autotune (possible for all Heaters) @Configuration->Temperatures->(T0 Tn Bed)->Method
- Added Menu for changing "PID drive min" and "PID drive max" (possible for all Heaters) @Configuration->Temperatures->(T0 Tn Bed)
- Added Menu for changing "PID max" (possible for all Heaters) @Configuration->Temperatures->(T0 Tn Bed)
- Added Menu for changing "Sensortype" (possible for all Heaters) @Configuration->Temperatures->(T0 Tn Bed)
- Renamed "PID drive min" max to "PID I-drive min" for better understanding
- Renamed "PID drive max" max to "PID I-drive min" for better understanding
- Renamed "PID max" to "Power Max" for better understanding and to prevent mixups with "PID drive max"
- Tiny Fix for constrain at autotune by Gcode.

V RF.01.37r6.Mod (2017-08-01)
- Emergency-Stop for Z-Moves is now configurable within EEPROM and Printers Menu
- Text for RGB-Light is now "Auto" instead Color anymore. 
- Cleaned up EEPROM text for RGB-Light to get better Compatibility with Repetier-Server (Value is not hidden anymore)
- New Standard PID-Parameters for V2 because the old ones are quite bad in context to the improved new PID control.

V RF.01.37r5.Mod (2017-07-31)
- "PID drive max = 40"-Fix.
- Removed FEATURE_CONFIGURABLE_HOTEND because it was bogus on Dual and not really missed at RF1000 Single setup.
- Extruder Sensor Type is now configurable within EEPROM
- Cleanup, deleted unnecessary text and defines
- Added Thermistor-Types description as Comment to RFx000.h configurations

V RF.01.37r4.Mod (2017-07-29)
- Bugfix for Z-Matrix-Invalidation
- Extruder 2 Z-Offset got eeprom support
- Cleanup

V RF.01.37r3.Mod (2017-07-24)
- Fix M3902 Z0 to update Z-Offset in Flash too.
- Fix min-max-bug in Emergency-Pause Menu

V RF.01.37r2.TDMod (2017-07-24)
- Last SenseOffset Digits Limit is now stored in EEPROM. If you use a plain M3909 instead of M3909 Pxxxxx the last value is used from EEPROM automatically.
- The Emergency-Pause limit values are now editable and stored in EEPROM. EEPROM values are loaded at the printers start and can be updated via M3075.
- Emergency-Pause is now editable within Menu at Configuration->General (See at V 0.91.15 (2014-03-26) vor Details.)
- Updated small things according to repetier/Repetier @output EEPROM and some cleanup.
- Updated PID D-Gains temperature-diff to 1500ms timespan. This smoothes out the steering.

V RF.01.37r.TDMod (2017-07-22)
- Grouped my good old Auto-Matrix-Leveling to one Command: M3900 R1. This is available in Menu also. Configuration -> Z-Configuration -> "Auto-Matrix-Leveling"
- Z-Step-Size which can be set in Positon-Menu is now saved within flash automatically.
- Previous used maximal SenseOffset-Offsetlimit is now saved within flash automatically.

Repetier/Constants.h
/*
    This file is part of the Repetier-Firmware for RF devices from Conrad Electronic SE.

    Repetier-Firmware is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    Repetier-Firmware is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with Repetier-Firmware.  If not, see <http://www.gnu.org/licenses/>.
*/


#ifndef CONSTANTS_H
#define CONSTANTS_H


#define REPETIER_VERSION                    "RF.01.37s9.Mod"
#define UI_PRINTER_COMPANY                  "DGLASS RF1k Edt."   //Digibike Change
#define UI_VERSION_STRING                   "V " REPETIER_VERSION


// ##########################################################################################
// ##    basic definitions
// ##########################################################################################

#define DEVICE_TYPE_RF1000                  13
#define DEVICE_TYPE_RF2000                  14

#define OPERATING_MODE_PRINT                1   // the firmware works in mode "print"
#define OPERATING_MODE_MILL                 2   // the firmware works in mode "mill"

#define HOTEND_TYPE_V1                      2   // hotend V1
#define HOTEND_TYPE_V2                      3   // hotend V2 for single extruder

#define MILLER_TYPE_ONE_TRACK               1   // one track in x- and y-direction
#define MILLER_TYPE_TWO_TRACKS              2   // two tracks in x- and y-direction

#define ENDSTOP_TYPE_SINGLE                 1   // there is only one endstop attached (either the min- or the max-endstop)
#define ENDSTOP_TYPE_CIRCUIT                2   // the min- and max-endstops are attached in a single circuit

#define ENDSTOP_NOT_HIT                     0
#define ENDSTOP_IS_HIT                      1
#define ENDSTOP_WAS_HIT                     2

#define X_AXIS                              0
#define Y_AXIS                              1
#define Z_AXIS                              2
#define E_AXIS                              3
#define VIRTUAL_AXIS                        4

#define HOME_ORDER_XYZ                      1
#define HOME_ORDER_XZY                      2
#define HOME_ORDER_YXZ                      3
#define HOME_ORDER_YZX                      4
#define HOME_ORDER_ZXY                      5
#define HOME_ORDER_ZYX                      6

#define MENU_MODE_SD_MOUNTED                1
#define MENU_MODE_SD_PRINTING               2
#define MENU_MODE_SD_PAUSED                 4
#define MENU_MODE_FAN_RUNNING               8
#define MENU_MODE_PRINTING                  16
#define MENU_MODE_PRINTER                   32  // we have to show the printer menu
#define MENU_MODE_MILLER                    64  // we have to show the miller menu

#define IGNORE_COORDINATE                   99999

#define TASK_NO_TASK                        0
#define TASK_ENABLE_Z_COMPENSATION          1
#define TASK_DISABLE_Z_COMPENSATION         2
#define TASK_PAUSE_PRINT                    4
#define TASK_PAUSE_PRINT_AND_MOVE           5
#define TASK_MOVE_FROM_BUTTON               10

#define PAUSE_STATUS_NONE                   0   // we are not paused at the moment
#define PAUSE_STATUS_PAUSED                 1   // we have paused the print (= pause position reached)

#define PAUSE_STATUS_GOTO_PAUSE1            2   // we are waiting until the last queue move is finished
#define PAUSE_STATUS_GOTO_PAUSE2            3   // we are going to pause the print/milling (= moving to the pause position)
#define PAUSE_STATUS_GOTO_PAUSE3            4   // we are going to pause the milling second stage: moving Z
#define PAUSE_STATUS_PREPARE_CONTINUE2_1    5   // we are going to continue the print (= moving to the continue position)
#define PAUSE_STATUS_PREPARE_CONTINUE2_2    6   // we are going to continue the print (= moving to the continue position)
#define PAUSE_STATUS_PREPARE_CONTINUE1      7   // we are going to continue the print (= moving to the continue position)

#define PAUSE_STATUS_TASKGOTO_PAUSE_1       7   // we are going to pause the print (= moving to the pause position)
#define PAUSE_STATUS_TASKGOTO_PAUSE_2       8   // we are going to pause the print (= moving to the pause position)
#define PAUSE_STATUS_TASKGOTO_PAUSE_3       9   // we are going to pause the print (= moving to the pause position)

#define PAUSE_MODE_NONE                     0   // we are not paused at the moment
#define PAUSE_MODE_PAUSED                   1   // stopp at the last printing position
#define PAUSE_MODE_PAUSED_AND_MOVED         2   // move away from the last printing position

#define RGB_MODE_OFF                        0
#define RGB_MODE_WHITE                      1
#define RGB_MODE_AUTOMATIC                  2
#define RGB_MODE_MANUAL                     3

#define RGB_STATUS_NOT_AUTOMATIC            0
#define RGB_STATUS_AUTOMATIC                1
#define RGB_STATUS_PRINTING                 11
#define RGB_STATUS_HEATING                  12
#define RGB_STATUS_COOLING                  13
#define RGB_STATUS_IDLE                     14
#define RGB_STATUS_COLOR_CHANGE             15

#define MOVE_MODE_SINGLE_STEPS              1
#define MOVE_MODE_SINGLE_MOVE               2
#define MOVE_MODE_1_MM                      3
#define MOVE_MODE_10_MM                     4
#define MOVE_MODE_50_MM                     5

#define HEAT_BED_SCAN_MODE_PLA              1
#define HEAT_BED_SCAN_MODE_ABS              2

#define Z_VALUE_MODE_Z_MIN                  1   // show the z-distance to z-min (print)
#define Z_VALUE_MODE_Z_ORIGIN               1   // show the z-distance to the z-origin (mill)
#define Z_VALUE_MODE_SURFACE                2   // show the z-distance to the surface of the heat bed (print) or work part (mill)
#define Z_VALUE_MODE_LAYER                  3   // show the z-distance to the surface of the heat bed (print) or work part (mill)

#define STOP_BECAUSE_OF_Z_MIN               1
#define STOP_BECAUSE_OF_Z_BLOCK             2


// ##########################################################################################
// ##    data types
// ##########################################################################################

#define uint                                uint16_t
#define uint8                               uint8_t
#define int8                                int8_t
#define uint32                              uint32_t
#define int32                               int32_t


#endif // CONSTANTS_H
